controllers/replace_misconfigured_process_groups.go

```diff
@@ -34,67 +34,69 @@ import (
 	corev1 "k8s.io/api/core/v1"
 )
 
-// ReplaceMisconfiguredPods identifies processes that need to be replaced in
+// ReplaceMisconfiguredProcessGroups identifies processes that need to be replaced in
 // order to bring up new processes with different configuration.
-type ReplaceMisconfiguredPods struct{}
+type ReplaceMisconfiguredProcessGroups struct{}
 
 // Reconcile runs the reconciler's work.
-func (c ReplaceMisconfiguredPods) Reconcile(r *FoundationDBClusterReconciler, context ctx.Context, cluster *fdbtypes.FoundationDBCluster) *Requeue {
+func (c ReplaceMisconfiguredProcessGroups) Reconcile(r *FoundationDBClusterReconciler, context ctx.Context, cluster *fdbtypes.FoundationDBCluster) *Requeue {
+	logger := log.WithValues("namespace", cluster.Namespace, "cluster", cluster.Name, "reconciler", "ReplaceMisconfiguredProcessGroups")
 	hasNewRemovals := false
 
-	processGroups := make(map[string]*fdbtypes.ProcessGroupStatus)
-	for _, processGroup := range cluster.Status.ProcessGroups {
-		processGroups[processGroup.ProcessGroupID] = processGroup
-	}
-
 	pvcs := &corev1.PersistentVolumeClaimList{}
 	err := r.List(context, pvcs, internal.GetPodListOptions(cluster, "", "")...)
 	if err != nil {
 		return &Requeue{Error: err}
 	}
 
-	for _, pvc := range pvcs.Items {
-		instanceID := internal.GetProcessGroupIDFromMeta(cluster, pvc.ObjectMeta)
-		processGroupStatus := processGroups[instanceID]
-		if processGroupStatus == nil {
-			return &Requeue{Error: fmt.Errorf("unknown PVC %s in replace_misconfigured_pods", instanceID)}
-		}
+	pvcMap := internal.CreatePVCMap(cluster, pvcs)
+
+	pods, err := r.PodLifecycleManager.GetPods(r, cluster, context, internal.GetPodListOptions(cluster, "", "")...)
+	if err != nil {
+		return &Requeue{Error: err}
+	}
 
-		if processGroupStatus.Remove {
+	podMap := internal.CreatePodMap(cluster, pods)
+
+	for _, processGroup := range cluster.Status.ProcessGroups {
+		if processGroup.Remove {
 			continue
 		}
 
-		needsNewRemoval, err := instanceNeedsRemovalForPVC(cluster, pvc)
-		if err != nil {
-			return &Requeue{Error: err}
-		}
-		if needsNewRemoval {
-			instances, err := r.PodLifecycleManager.GetPods(r, cluster, context, internal.GetSinglePodListOptions(cluster, instanceID)...)
+		pvc, hasPVC := pvcMap[processGroup.ProcessGroupID]
+		pod, hasPod := podMap[processGroup.ProcessGroupID]
+
+		if hasPVC {
+			needsPVCRemoval, err := instanceNeedsRemovalForPVC(cluster, pvc)
 			if err != nil {
 				return &Requeue{Error: err}
 			}
 
-			if len(instances) > 0 {
-				processGroupStatus.Remove = true
-				hasNewRemovals = true
+			if needsPVCRemoval {
+				if hasPod {
```

Comment:
```suggestion
			if needsPVCRemoval && hasPod {
```

I'm actually not sure if we really need to check the `hasPod` condition. I rather think we can drop it.